### PR TITLE
Fix deployment helper - no assumptions on only one new ReplicaSet

### LINKED ISSUE #40415: Upgrading to v1.5.2 creates duplicate replica sets
**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.): multiple replicaset, duplicate replicaset

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): **BUG REPORT**

**Kubernetes version** (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.2", GitCommit:"08e099554f3c31f6e6f07b448ab3ed78d0520507", GitTreeState:"clean", BuildDate:"2017-01-12T04:57:25Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.2", GitCommit:"08e099554f3c31f6e6f07b448ab3ed78d0520507", GitTreeState:"clean", BuildDate:"2017-01-12T04:52:34Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}
```

**Environment**:
- **Cloud provider or hardware configuration**: **AWS**
- **OS** (e.g. from /etc/os-release): **Ubuntu 14.04.3 LTS**
- **Kernel** (e.g. `uname -a`): **Linux ip-10-0-10-162 3.13.0-87-generic #133-Ubuntu SMP Tue May 24 18:32:09 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux**
- **Install tools**: Custom terraform+ansible setup; using hyperkube binary
- **Others**: This cluster has disabled auth and serviceaccounts

**What happened**: Upgrading control plane to v1.5.2 (from v1.4.7) created duplicate replicasets (and hence duplicate pods) for some deployments

**What you expected to happen**: Existing deployments/replicasets/pods to be the same OR being replaced - not duplicates

**How to reproduce it** (as minimally and precisely as possible): Keep a deployment with volume mounts (probably EBS) running on v1.4.7. Upgrade control plane components to v1.5.2

**Anything else do we need to know**: Interestingly, this seems to have happened only for deployments that use shared volume mounts (either EBS or otherwise).

### PR DESCRIPTION
**Pull request by janetkuo**
#40415

**Release note**:

```release-note
NONE
```

@kubernetes/sig-apps-bugs

### CODE REVIEW

**Comment by kargakis** [comment]
/lgtm

We should cherry-pick this to 1.5

**Comment by janetkuo** [comment]
@k8s-bot non-cri e2e test this
@k8s-bot gci gce e2e test this
@k8s-bot gce etcd3 e2e test this

**Comment by janetkuo** [comment]
cc @mwielgus for cherrypick approval

**Comment by kargakis** [comment]
@k8s-bot gce etcd3 e2e test this

File: pkg/controller/deployment/util/deployment_util.go [hunk 1]
```diff
@@ -612,9 +612,10 @@ func EqualIgnoreHash(template1, template2 v1.PodTemplateSpec) bool {
 // FindNewReplicaSet returns the new RS this given deployment targets (the one with the same pod template).
 func FindNewReplicaSet(deployment *extensions.Deployment, rsList []*extensions.ReplicaSet) (*extensions.ReplicaSet, error) {
 	newRSTemplate := GetNewReplicaSetTemplate(deployment)
+	sort.Sort(controller.ReplicaSetsByCreationTimestamp(rsList))
```
Comment (kargakis) [inline pkg/controller/deployment/util/deployment_util.go]: @smarterclayton @mfojtik we are going to choose the oldest replica set out of two identical replica sets for now - seems like the sanest thing to do when we end up in such a state (seems possible after cluster upgrades: https://github.com/kubernetes/kubernetes/issues/40415).

File: pkg/controller/deployment/util/deployment_util.go [hunk 2]
```diff
@@ -629,16 +630,21 @@ func FindOldReplicaSets(deployment *extensions.Deployment, rsList []*extensions.
...
 		podLabelsSelector := labels.Set(pod.ObjectMeta.Labels)
 		for _, rs := range rsList {
 			rsLabelsSelector, err := metav1.LabelSelectorAsSelector(rs.Spec.Selector)
 			if err != nil {
 				return nil, nil, fmt.Errorf("invalid label selector: %v", err)
 			}
-			// Filter out replica set that has the same pod template spec as the deployment - that is the new replica set.
-			if EqualIgnoreHash(rs.Spec.Template, newRSTemplate) {
+			// Filter out new replica set
+			if rs == nil || (newRS != nil && rs.UID == newRS.UID) {
```
Comment (kargakis) [inline pkg/controller/deployment/util/deployment_util.go]: rs cannot be nil here.

File: pkg/controller/deployment/util/deployment_util.go (hunk 1 above)
Comment (kargakis) [inline pkg/controller/deployment/util/deployment_util.go]: @janetkuo can you add a comment explaining this here?

File: pkg/controller/deployment/util/deployment_util.go [hunk 3]
```diff
@@ -629,16 +630,21 @@ func FindOldReplicaSets(deployment *extensions.Deployment, rsList []*extensions.
...
 	oldRSs := map[string]*extensions.ReplicaSet{}
 	allOldRSs := map[string]*extensions.ReplicaSet{}
-	newRSTemplate := GetNewReplicaSetTemplate(deployment)
+	requiredRSs := []*extensions.ReplicaSet{}
+	allRSs := []*extensions.ReplicaSet{}
+	newRS, err := FindNewReplicaSet(deployment, rsList)
+	if err != nil {
+		return requiredRSs, allRSs, err
+	}
 	for _, pod := range podList.Items {
```
Comment (kargakis) [inline pkg/controller/deployment/util/deployment_util.go]: I have PR that reworks `FindOldReplicaSets` - the podList is not needed at all. We should simply look at rs.spec.replicas for filtering RSs with and without pods.

**Comment by kargakis** [comment]
@janetkuo just two comments, reapply lgtm once they are fixed.

**Comment by janetkuo** [comment]
Test failed because #41650 added staging godeps check. Running `hack/update-staging-godeps.sh` to fix it.

File: pkg/controller/deployment/util/deployment_util.go (hunk 1 above)
Comment (janetkuo) [inline pkg/controller/deployment/util/deployment_util.go]: done

**Comment by smarterclayton** [comment]
I just saw https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/41729/pull-kubernetes-bazel/6787/ flake - is this related to the changes made here?

**Comment by kargakis** [comment]
> I just saw https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/41729/pull-kubernetes-bazel/6787/ flake - is this related to the changes made here?

Yes, @janetkuo already opened a fix #41961

**Pull request merged**